src/PureSMT/Process.hs

```diff
@@ -1,74 +1,67 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
 module PureSMT.Process where
 
 import Control.Monad
-import Data.Maybe (fromMaybe)
-import Data.String (fromString)
+import qualified Data.ByteString.Char8 as BS
+import Foreign.Ptr (Ptr)
+import qualified Language.C.Inline as C hiding (block, exp)
+import qualified Language.C.Inline.Unsafe as C
 import PureSMT.SExpr
+import qualified PureSMT.Z3 as Z3
 import System.IO
-import System.Mem.Weak
-import qualified System.Process as P
-import System.Process.Typed
 import Prelude hiding (const)
 
-type SolverProcess = Process Handle Handle Handle
+C.context (C.baseCtx <> C.bsCtx <> Z3.cContext)
+C.include "z3.h"
 
 data Solver = Solver
-  { process :: SolverProcess,
+  { context :: Ptr Z3.LogicalContext,
     debugMode :: Bool
   }
 
-solverPid :: Solver -> IO (Maybe P.Pid)
-solverPid = P.getPid . unsafeProcessHandle . process
-
-unsafeSolverPid :: Solver -> IO P.Pid
-unsafeSolverPid = fmap (fromMaybe $ error "ProcessHandle was already closed") . solverPid
-
-launchSolverWithFinalizer ::
-  -- | Command to call the solver, will be made into a 'ProcessConfig' with 'fromString'
-  String ->
+-- | Create a brand-new context for Z3 to work in.
+initZ3instance ::
   -- | Whether or not to debug the interaction
   Bool ->
   IO Solver
-launchSolverWithFinalizer cmd dbg = do
-  p <- startProcess config
-
-  let s = Solver p dbg
+initZ3instance dbg = do
+  solverCtx <-
+    [C.block| Z3_context {
+                     Z3_config cfg = Z3_mk_config();
+                     Z3_context ctx = Z3_mk_context(cfg);
+                     Z3_del_config(cfg);
+                     return ctx;
+                     } |]
+
+  let s = Solver solverCtx dbg
   setOption s ":print-success" "true"
   setOption s ":produce-models" "true"
 
-  -- Registers a finalizer to send an "exit" command to the solver, in case it
-  -- hasn't terminated yet. If it has terminated, we do nothing.
-  addFinalizer p (getExitCode p >>= maybe (send s (List [Atom "exit"])) (\_ -> return ()))
   return s
-  where
-    config :: ProcessConfig Handle Handle Handle
-    config = setStdin createPipe $ setStdout createPipe $ setStderr createPipe $ fromString cmd
 
-send :: Solver -> SExpr -> IO ()
-send solver cmd = do
-  let cmdTxt = showsSExpr cmd ""
+-- | Have Z3 evaluate a command in SExpr format.
+command :: Solver -> SExpr -> IO SExpr
+command solver cmd = do
+  let cmdTxt = serializeSExpr cmd
   when (debugMode solver) $ do
-    pid <- unsafeSolverPid solver
-    putStrLn ("[send: " ++ show pid ++ "] " ++ cmdTxt)
-  hPutStrLn (getStdin $ process solver) cmdTxt
-  hFlush (getStdin $ process solver)
-
-recv :: Solver -> IO SExpr
-recv solver = do
-  resp <- hGetLine (getStdout $ process solver)
+    BS.putStrLn $ "[send] " `BS.append` cmdTxt
+  let ctx = context solver
+  resp <-
+    [C.exp| const char* {
+                  Z3_eval_smtlib2_string($(Z3_context ctx), $bs-cstr:cmdTxt)
+                  } |]
+      >>= BS.packCString
```

Comment:
This is leaking the output bytestring after packing it, right?

In https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Unsafe.html#g:3 there are functions unsafePackCString and unsafePackCStringFinalizer that could be helpful.

Reply:
Yeah indeed I hadn't thought of that. So I tried doing 
```
[CU.exp| char* {
                  // the function outputs a const char*
                  (char*) Z3_eval_smtlib2_string($(Z3_context ctx), $bs-cstr:cmdTxt)
                  } |]
>>= BS.unsafePackMallocCString
```
because calling `free` on the CString is the only finalizer I can think of (Z3 doesn't have a `del_string` function in its API because `Z3_string`s are just `const char*`s) and you can't free `const`s. But the tests fail at run-time with
```
  Symbolic
    Symbolic prove
      incorrectness triples
        [input > 0] add 1 [result > 0] counter:                                                           OK (0.01s)
        [input > 0] add 1 [result > 1] verified:                                                          free(): invalid size
./tests/run-tests.sh: line 57: 1297697 Aborted                 (core dumped) cabal run tests
[FAILURE] 'cabal run tests' failed; check the respective artifact.
```
I assume the `free() : invalid size` means it's trying to free a non-allocated pointer, but I don't see how it could happen when the first example works perfectly fine.

Reply:
Also if we're worried about memory leaks then the context also gets leaked doesn't it ? We don't send `(exit)` anymore so Z3 can't possibly know it should call `Z3_del_context`.

Reply:
> Also if we're worried about memory leaks then the context also gets leaked doesn't it ?

That's a good point. In the case of the context, I'm guessing it is created only once per execution. But if pirouette were used as a library, perhaps it could be leaked multiple times.

The queries to the SMT solver do happen enough times that the leaks could be problematic in sufficiently large examples.

> but I don't see how it could happen when the first example works perfectly fine.

I went to check how the `Z3_string` is created. `Z3_eval_smtlib2_string` calls in turn to a function `mk_external_string` which returns a reference to a buffer kept in the context. If the reference is released, most likely the next call to Z3 that tries to use the buffer is going to fail.

In short, it looks like Z3 doesn't expect us to release the `Z3_string`s. I also checked [one of the examples](https://github.com/Z3Prover/z3/blob/fe1e3017354768e7761ad1db0805fb54b5d4f333/examples/c/test_capi.c) and it doesn't release `Z3_string`s either. So I'm a bit puzzled, how can we have more than one `Z3_string` at once if they both are pointers to the same buffer? There seem to be API calls which take more than one `Z3_string` as parameter.

Reply:
So in the end I should use `unsafePackCString` instead of `packCString` ? In both cases the underlying CString is not garbage collected but in the safe case the CString is copied to create the bytestring so it should be slower than the unsafe case which just computes the length of the string, right ?

Reply:
Given that a later call to `mk_external_string` could invalidate previously returned Z3 strings, I think the safest option is to immediately copy the CString as your original code does.

Reply:
Given that the output bytestring is parsed and consumed (by `readSExpr`) inside the body of `command`, the unsafeness will only lead to an error if two instances of `command` are run concurrently and with the same solver instance (I'm assuming all of Pirouette's interaction with Z3 is done inside `command`). But Z3 is only considered thread-safe if the different threads use different contexts so I think it makes sense to keep this.

Reply:
> Given that the output bytestring is parsed and consumed (by readSExpr) inside the body of command

I would agree with that, if `command` deepseq'ed the resulting SExp before returning it to the caller. As it stands, some suspended computation in the strings inside the SExp could leak a reference to the bytestring.

Reply:
Also, I would hope that copying the bytestring is faster than deepseq'ing, but you might disprove me.

Reply:
Oh yeah I didn't consider that. It would still be fine if the output SExpr were consumed before sending the next command but that wouldn't be compatible with exploring several branches in parallel.
And well I would guess copying bytestrings would make the overall amount of work bigger than just deepseq'ing, but that would only be true under the assumption that Pirouette always uses the output SExpr and I imagine it's not necessary when the command is something that will just return `success`. I'll just revert to `packCString` then.